src/iaphandler.cpp

```diff
@@ -0,0 +1,282 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+#include "iaphandler.h"
+#include "constants.h"
+#include "leakdetector.h"
+#include "logger.h"
+
+#include <QCoreApplication>
+#include <QJsonArray>
+#include <QJsonDocument>
+#include <QJsonObject>
+#include <QJsonValue>
+#include <QScopeGuard>
+
+#ifdef MVPN_IOS
+#  include "platforms/ios/iosiaphandler.h"
+#else
+#  include "platforms/dummy/dummyiaphandler.h"
+#endif
+
+namespace {
+Logger logger(LOG_IAP, "IAPHandler");
+}  // namespace
+
+// static
+IAPHandler* IAPHandler::createInstance() {
+#ifdef MVPN_IOS
+  return IOSIAPHandler::createInstance();
+#else
+  return DummyIAPHandler::createInstance();
+#endif
+}
+
+// static
+IAPHandler* IAPHandler::instance() {
+#ifdef MVPN_IOS
+  return IOSIAPHandler::instance();
+#else
+  return DummyIAPHandler::instance();
+#endif
+}
+
+void IAPHandler::registerProducts(const QByteArray& data) {
+  logger.debug() << "Maybe register products";
+
+  Q_ASSERT(m_productsRegistrationState == eRegistered ||
+           m_productsRegistrationState == eNotRegistered);
+
+  auto guard = qScopeGuard([&] { emit productsRegistered(); });
+
+  if (m_productsRegistrationState == eRegistered) {
+    return;
+  }
+
+  Q_ASSERT(m_products.isEmpty());
+
+  QJsonDocument json = QJsonDocument::fromJson(data);
+  if (!json.isObject()) {
+    logger.debug() << "Object expected";
+    return;
+  }
+
+  QJsonObject obj = json.object();
+  if (!obj.contains("products")) {
+    logger.debug() << "products entry expected";
+    return;
+  }
+
+  QJsonArray products = obj["products"].toArray();
+  if (products.isEmpty()) {
+    logger.error() << "No products found";
+    return;
+  }
+
+  m_productsRegistrationState = eRegistering;
+
+  for (auto p : products) {
```

Comment:
see comment in iosiaphandler.mm

Reply:
This is OK! You can also write: `QJsonValue p : product`